### PR TITLE
`<mutex>`: Scope guards for lock algorithms and exception specification strengthening

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
This PR changes `_RERAISE` in `<mutex>` to scope guards (towards #2307).
And strengthens exception specifications for
- `_Mutex_base::unlock` (i.e. `mutex::unlock` and `recursive_mutex::unlock`), and
- move assignment operator of `unique_lock` ("reverting LWG-2104", addressing #3278).

### CODE REVIEW

File: stl/inc/mutex
Comment (strega-nil-ms) [inline stl/inc/mutex:63]: I'm noticing that `try_lock` is not similarly marked, and in fact does throw sometimes, despite having the same requirement of "_Throws_: Nothing" in the standard. I am concerned that this might be a bug in our implementation.

File: stl/inc/mutex
Comment (strega-nil-ms) [inline stl/inc/mutex:303]: I do not love this name. I'd prefer something along the lines of `_Call_guard`, since that's what it does - it calls the function object passed in.

File: stl/inc/mutex
Comment (strega-nil-ms) [inline stl/inc/mutex:304]:
```suggestion
    static_assert(is_trivially_copyable_v<_Fn>, "This scope guard is only used for trivially copyable function objects.");
```

personal pet peeve - functor is already an insanely overloaded term, we should not additionally add the meaning of "function object".

Additionally, below, `_Functor` -> `_Fx` or `_Func`; `_Ftor` -> `_Func`.

Comment (strega-nil-ms) [reply stl/inc/mutex:304]: I also don't really understand why the function object being trivially copyable is important.

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

File: stl/inc/mutex
Comment (CaseyCarter) [reply stl/inc/mutex:304]: I suspect that "trivially copyable" is being used here because it implies `is_nothrow_copy_constructible_v`. Absent that requirement, we'd need a `try` block in the constructor to ensure the function object is called if copying it throws.

File: stl/inc/mutex
Comment (CaseyCarter) [reply stl/inc/mutex:63]: I suspect that `try_lock` only throws on precondition violation, which would allow us to strengthen, but we need to validate that expectation.

File: stl/inc/mutex
Comment (frederick-vs-ja) [reply stl/inc/mutex]: I wanted to make a stronger assertion that initialization of the guard is not only non-throwing, but also has no side effects...

**Review by strega-nil-ms — APPROVED** [review]
Thanks so much!

File: stl/inc/mutex
Comment (CaseyCarter) [reply stl/inc/mutex:63]: I've filed #3324 to investigate the `try_lock` issue so it need not hold up this PR.

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/mutex
Comment (CaseyCarter) [inline stl/inc/mutex:307]:
```suggestion
    explicit _Unlock_call_guard(const _Fn& _Fx) noexcept : _Func(_Fx) {}
```

File: stl/inc/mutex
Comment (CaseyCarter) [inline stl/inc/mutex:324]:
```suggestion
    explicit _Unlock_one_guard(_Lock& _Lk) noexcept : _Lk_ptr(_STD addressof(_Lk)) {}
```

**Review by CaseyCarter — APPROVED** [review]
I lied: found something to change =)

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, these transformations look exactly correct, and bring us one step closer to the goal of eliminating unnecessary `_RERAISE`! :heart_eyes_cat:

FYI @CaseyCarter @strega-nil-ms I added a couple of `_NODISCARD`s after you approved.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks again for these nontrivial changes, improving debug codegen and debuggability! :white_check_mark: :smile_cat: :tada:

**Pull request merged**
